crud/common/sharding_key.lua

```diff
@@ -0,0 +1,122 @@
+local fiber = require('fiber')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local const = require('crud.common.const')
+local dev_checks = require('crud.common.dev_checks')
+local cache = require('crud.common.sharding_key_cache')
+local utils = require('crud.common.utils')
+
+local FetchShardingKeyError = errors.new_class('FetchShardingKeyError', {capture_stack = false})
+local GetShardingKeyFieldnosError = errors.new_class('GetShardingKeyFieldnosError',  {capture_stack = false})
+
+local FETCH_FUNC_NAME = '_crud.fetch_on_storage'
+
+local sharding_key_module = {}
+
+-- Build a structure similar to index, but it is not a real index object,
+-- it contains only parts key with fieldno's.
+function sharding_key_module.as_index_object(space_format, space_name, sharding_key_def)
+    dev_checks('table', 'string', 'table')
+
+    local fieldnos = {}
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    for _, field_name in ipairs(sharding_key_def) do
+        local fieldno = fieldno_map[field_name]
+        if fieldno == nil then
+            return nil, GetShardingKeyFieldnosError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- It's a public method to access it in tests.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_def = {}
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return sharding_key_def
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    for _, tuple in ipairs(sharding_key_space:select()) do
```

Comment:
Now this code always work locally (on `box`, not a net.box connection), so `<space>:pairs()` would look better than `ipairs(<space>:select())`. If it'll be changed in next commits (I didn't read the whole patchset yet), okay, leave it as is.

Reply:
Now switched to `pairs()`:

```lua
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -44,7 +44,7 @@ function sharding_key_module.fetch_on_storage()
 
     local SPACE_NAME_FIELDNO = 1
     local SPACE_SHARDING_KEY_FIELDNO = 2
-    for _, tuple in ipairs(sharding_key_space:select()) do
+    for _, tuple in sharding_key_space:pairs() do
         local space_name = tuple[SPACE_NAME_FIELDNO]
         local key = tuple[SPACE_SHARDING_KEY_FIELDNO]
         sharding_key_def[space_name] = key
```